tests/test_qdrant_client.py

```diff
@@ -2305,6 +2309,204 @@ def test_strict_mode(prefer_grpc):
+    if client.collection_exists(COLLECTION_NAME):
+        client.delete_collection(collection_name=COLLECTION_NAME, timeout=TIMEOUT)
+    client.create_collection(
+        collection_name=COLLECTION_NAME,
+        vectors_config=models.VectorParams(size=DIM, distance=models.Distance.DOT),
+        timeout=TIMEOUT,
+    )
+
+    points = generate_points(num_points=100, vector_sizes=DIM)
+    ids = []
+    vectors = []
+    for point in points:
+        ids.append(point.id)
+        vectors.append(point.vector)
+
+    points_batch = models.Batch(
+        ids=ids,
+        vectors=vectors,
+        payloads=None,
+    )
+
+    client.update_collection(
+        collection_name=COLLECTION_NAME,
+        strict_mode_config=models.StrictModeConfig(
+            enabled=True, read_rate_limit=READ_LIMIT, write_rate_limit=WRITE_LIMIT
+        ),
+    )
+
+    if prefer_grpc:
+        with pytest.raises(
+            RpcError,
+            match="Write rate limit exceeded",
+        ):
+            client.upsert(COLLECTION_NAME, points_batch)
+    else:
+        with pytest.raises(
+            qdrant_client.http.exceptions.UnexpectedResponse,
+            match="Write rate limit exceeded",
+        ):
+            client.upsert(COLLECTION_NAME, points_batch)
+
+
+@pytest.mark.parametrize("prefer_grpc", [False, True])
+def test_upsert_hits_write_rate_limit(prefer_grpc):
+    client = QdrantClient(prefer_grpc=prefer_grpc, timeout=TIMEOUT)
+    if client.collection_exists(COLLECTION_NAME):
+        client.delete_collection(collection_name=COLLECTION_NAME, timeout=TIMEOUT)
+    client.create_collection(
+        collection_name=COLLECTION_NAME,
+        vectors_config=models.VectorParams(size=DIM, distance=models.Distance.DOT),
+        timeout=TIMEOUT,
+    )
+
+    points = generate_points(num_points=WRITE_LIMIT, vector_sizes=DIM)
+    ids = []
+    vectors = []
+    for point in points:
+        ids.append(point.id)
+        vectors.append(point.vector)
+
+    points_batch = models.Batch(
+        ids=ids,
+        vectors=vectors,
+        payloads=None,
+    )
+
+    client.update_collection(
+        collection_name=COLLECTION_NAME,
+        strict_mode_config=models.StrictModeConfig(
+            enabled=True, read_rate_limit=READ_LIMIT, write_rate_limit=WRITE_LIMIT
+        ),
+    )
+
+    flag = False
+    time_to_sleep = 0
+    try:
+        for _ in range(10):
+            client.upsert(collection_name=COLLECTION_NAME, points=points_batch)
+    except ResourceExhaustedResponse as ex:
+        print(ex.message)
+        assert ex.retry_after_s > 0, f"Unexpected retry_after_s value: {ex.retry_after_s}"
+        flag = True
+        time_to_sleep = int(ex.retry_after_s)
+
+    if flag:
+        # verify next response after sleep succeeds
+        sleep(time_to_sleep)
+        client.upsert(collection_name=COLLECTION_NAME, points=points_batch)
+    else:
+        raise AssertionError("No ResourceExhaustedResponse exception was raised.")
```

Comment:
let's not wait for all the points to upload but just check the exception
otherwise this test takes ~2 min, which is too long